Wallet:  Add `maxfeerate` wallet startup option

ismaelsadeeq opened the pull request:
This PR fixes #29220

- The PR adds a wallet `-maxfeerate` startup option, as the upper limit of wallet transactions fee rate.

- This fixes the ambiguity of using `maxtxfee` value to check the upper limit of transactions fee rate.

-  Wallet will not create a transaction with fee rate above `maxfeerate` value. 

- This PR adds a functional test that ensure the behavior is enforced.

glozow reviewed (commented):
I mentioned adding a `maxfeerate` option in #29220, but I was more imagining it as a wallet-only option. I don't think that wallet configurations should bleed into how the node RPCs function. Instead, the interaction there should be for wallet to pass it as a param to `BroadcastTransaction` when the it submits a tx to the node.

I'm also not sure about `maxburnamount` being a node-wide config (https://github.com/bitcoin/bitcoin/issues/29217#issuecomment-1900585661)

ismaelsadeeq commented:
Thanks @glozow for your review.
Will put this PR in draft while addressing Approach feedback

luke-jr commented:
I agree this feels more like it should be a wallet option.

ismaelsadeeq committed:
[wallet]: enforce `-maxfeerate` on wallet transactions

This commit prevents the wallet from submitting or broadcasting transactions
above `-maxfeerate`, and tests the new functionality.

Compare wallet transaction fees against `maxfeerate.GetFee(vsize)` instead of
reconstructing a fee rate from the rounded fee. This allows transactions whose
requested fee rate is at the configured limit, even when the resulting fee is
rounded up to whole satoshis.

`rpc_psbt.py` functional test nodes are modified to start with a custom
`-maxfeerate=1`, because the test requires creating and broadcasting a
transaction with a high fee rate.

A node in some subtests in `wallet_fundrawtransaction.py` is restarted with
`-maxfeerate=1` because those subtests require the node's wallet to be able to
create and broadcast transactions with a high fee rate.

ismaelsadeeq commented on src/init.cpp:
```diff
@@ -624,7 +625,10 @@ void SetupServerArgs(ArgsManager& argsman)
-    argsman.AddArg("-incrementalrelayfee=<amt>", strprintf("Fee rate (in %s/kvB) used to define cost of relay, used for mempool limiting and replacement policy. (default: %s)", CURRENCY_UNIT, FormatMoney(DEFAULT_INCREMENTAL_RELAY_FEE)), ArgsManager::ALLOW_ANY | ArgsManager::DEBUG_ONLY, OptionsCategory::NODE_RELAY);
+    argsman.AddArg("-maxfeerate=<amt>", strprintf("Reject transactions whose fee rate is higher than the specified value, expressed in %s/kvB. Set to 0 to accept any fee rate. (default: %s %s/kvB)", CURRENCY_UNIT, FormatMoney(node::DEFAULT_MAX_RAW_TX_FEE_RATE.GetFeePerK()), CURRENCY_UNIT), ArgsManager::ALLOW_ANY, OptionsCategory::DEBUG_TEST);
+    argsman.AddArg("-maxburnamount=<amt>", strprintf("Reject transactions with provably unspendable outputs (e.g. 'datacarrier' outputs that use the OP_RETURN opcode) greater than the specified value, expressed in %s, if burning funds through unspendable outputs is desired, increase this value. This check is based on heuristics and does not guarantee spendability of outputs (default: 0%s).", CURR ...
+    argsman.AddArg("-incrementalrelayfee=<amt>", strprintf("Fee rate (in %s/kvB) used to define cost of relay, used for mempool limiting and replacement policy. (default: %s) If burning funds through unspendable outputs is desired, increase this value. This check is based on heuristics and does not guarantee spendability of outputs.",
```
This is gone now.

ismaelsadeeq commented on src/rpc/mempool.cpp:
```diff
@@ -43,13 +44,6 @@ static RPCHelpMan sendrawtransaction()
-            {"maxburnamount", RPCArg::Type::AMOUNT, RPCArg::Default{FormatMoney(0)},
-             "Reject transactions with provably unspendable outputs (e.g. 'datacarrier' outputs that use the OP_RETURN opcode) greater than the specified value, expressed in " + CURRENCY_UNIT + ".\n"
-             "If burning funds through unspendable outputs is desired, increase this value.\n"
-             "This check is based on heuristics and does not guarantee spendability of outputs.\n"},
```
Reverted thank you.

ismaelsadeeq commented on src/wallet/wallet.h:
```diff
@@ -729,6 +729,8 @@ class CWallet final : public WalletStorage, public interfaces::Chain::Notificati
     CAmount m_default_max_tx_fee{DEFAULT_TRANSACTION_MAXFEE};
 
+    /** Maximum transaction fee rate (in satoshis) used by default for the wallet */
+    CAmount m_default_max_tx_fee_rate{node::DEFAULT_MAX_RAW_TX_FEE_RATE.GetFeePerK()};
```
Done, `CFeeRate` is the right format since we are representing fee rate.

ismaelsadeeq replied:
>Shouldn't there be a check in CreateTransactionInternal that makes sure we don't make transactions above this feerate, and an arg passed to BroadcastTransaction? I also don't see a test case for it.

Agreed

I Added a check in `CreateTransactionInternal` in 0c9e0c5b2218c154b62275794b878af7e1457647. 

No need to pass an arg to `BroadcastTransaction` if the transaction fee rate is above `maxfeerate` it will not read `BroadcastTransaction` so it will just be a dead code.

I added a test for it.

josibake replied:
Interestingly, in https://github.com/bitcoin/bitcoin/blob/411ba32af21a56efa0a570b6aa8bf8f035410230/src/rpc/mempool.cpp#L89

`sendrawtransaction` takes a fee_rate argument, which is used to calculate a `max_tx_fee` based on the size of the transaction, and this `max_tx_fee` is passed to `BroadcastTransaction`. This means a user could typo an insane `maxfeerate` when calling `sendrawtransaction` that bypasses the default max tx fee.

Perhaps it does make more sense to have `BroadcastTransaction` take two arguments, one for `max_tx_fee` and another for `max_tx_fee_rate`, and check them independently. Probably out of scope for this PR, since this is focused on the wallet, but wanted to mention it as a potential follow-up.

ismaelsadeeq replied:
Thanks for highlighting this issue @josibake , I agree with your `BroadcastTransaction` should enforce check on both.
This will make `sendrawtransaction` returns a specific error message.


---
For this reason
`MAX_FEE_RATE_EXCEEDED` error string is now `Fee exceeds maximum configured by user.`
because if I add the option `maxtxfee` it will be misleading for in `sendrawtransaction` output

DanaBCannon replied:
Im Open For Business

ismaelsadeeq commented:
Forced pushed from https://github.com/bitcoin/bitcoin/commit/2576efcf689401dcbea85856c74efc64af0c2cb2 to https://github.com/bitcoin/bitcoin/commit/0c9e0c5b2218c154b62275794b878af7e1457647 Compare the [diff](https://github.com/bitcoin/bitcoin/compare/2576efcf689401dcbea85856c74efc64af0c2cb2..0c9e0c5b2218c154b62275794b878af7e1457647)
- All review comments are addressed
- Removed `maxburnamount` option
- Made `maxfeerate` wallet option
- Added a check in `CreateTransactionInternal` that makes sure we don't make transactions above this `-maxfeerate`.
- Fixed ambiguity between `-maxtxfee` and `maxfeerate`
- Prevent setting wallet feerate above `maxfeerate` from `settxfee` RPC
- Added a functional test case
- Updated PR OP

luke-jr commented on src/wallet/init.cpp:
```diff
@@ -62,6 +62,7 @@ void WalletInit::AddWalletOptions(ArgsManager& argsman) const
     argsman.AddArg("-maxapsfee=<n>", strprintf("Spend up to this amount in additional (absolute) fees (in %s) if it allows the use of partial spend avoidance (default: %s)", CURRENCY_UNIT, FormatMoney(DEFAULT_MAX_AVOIDPARTIALSPEND_FEE)), ArgsManager::ALLOW_ANY, OptionsCategory::WALLET);
     argsman.AddArg("-maxtxfee=<amt>", strprintf("Maximum total fees (in %s) to use in a single wallet transaction; setting this too low may abort large transactions (default: %s)",
         CURRENCY_UNIT, FormatMoney(DEFAULT_TRANSACTION_MAXFEE)), ArgsManager::ALLOW_ANY, OptionsCategory::DEBUG_TEST);
+    argsman.AddArg("-maxfeerate=<amt>", strprintf("Maximum fee rate (in %s) of a single wallet transaction (default: %s)", CURRENCY_UNIT, FormatMoney(DEFAULT_TRANSACTION_MAXFEERATE.GetFeePerK())), ArgsManager::ALLOW_ANY, OptionsCategory::WALLET);
```
This still isn't a wallet option...

ismaelsadeeq replied:
I dont understand why you said it is not a wallet option.
The `OptionsCategory` of this startup option is `WALLET`.
And its only used in the wallet.

Can you please expand on your comment.
Thanks

luke-jr replied:
Something set on each wallet, potentially different for each

ismaelsadeeq replied:
I opened a separate issue for this here https://github.com/bitcoin/bitcoin/issues/32462#issue-3052551984

glozow commented on test/functional/wallet_send.py:
```diff
@@ -184,6 +184,33 @@ def test_send(self, from_wallet, to_wallet=None, amount=None, data=None,
 
         return res
 
+    def test_maxfeerate(self):
```
Maybe add tests where you (successfully) configure the node's wallet using `-maxfeerate`?

glozow replied:
Similarly, it would be good to test that both maxfee and maxfeerate requirements must be met.

ismaelsadeeq replied:
> Maybe add tests where you (successfully) configure the node's wallet using -maxfeerate?

I dont understand what you mean by successfully configure nodes wallet.

The test you are commenting on configure node[0] with the `-maxfeerate` option.
Could you clarify your comment, thanks.

---
>  Similarly, it would be good to test that both maxfee and maxfeerate requirements must be met.

Their is a test for `maxtxfee` already, and I added one for `maxfeerate`.
https://github.com/bitcoin/bitcoin/blob/11b436a66af3ceaebb0f907878715f331516a0bc/test/functional/wallet_bumpfee.py#L557

glozow commented on src/wallet/wallet.h:
```diff
@@ -136,6 +136,8 @@ static const bool DEFAULT_DISABLE_WALLET = false;
 //! -maxtxfee default
 constexpr CAmount DEFAULT_TRANSACTION_MAXFEE{COIN / 10};
+//! maxfeerate default
+const CFeeRate DEFAULT_TRANSACTION_MAXFEERATE(CFeeRate((COIN / 10), 1000));
```
1000 is redundant

ismaelsadeeq replied:
Yes, fixed here and other place I did the same.

josibake commented on src/wallet/rpc/spend.cpp:
```diff
@@ -430,15 +430,14 @@ RPCHelpMan settxfee()
         throw JSONRPCError(RPC_INVALID_PARAMETER, strprintf("txfee cannot be less than wallet min fee (%s)", pwallet->m_min_fee.ToString()));
-    } else if (tx_fee_rate > max_tx_fee_rate) {
-        throw JSONRPCError(RPC_INVALID_PARAMETER, strprintf("txfee cannot be more than wallet max tx fee (%s)", max_tx_fee_rate.ToString()));
+    } else if (tx_fee_rate > pwallet->m_max_tx_fee_rate) {
+        throw JSONRPCError(RPC_INVALID_PARAMETER, strprintf("txfee cannot be more than max tx fee rate (%s)", pwallet->m_max_tx_fee_rate.ToString()));
```
I think we should keep the word "wallet" here, since this is a client set option. If you look at the RPC call above, it differentiates between the node wide policy of `relayMinFee` and the wallet set policy `m_min_fee` by adding the world "wallet" to the RPC error.

josibake commented on src/wallet/wallet.cpp:
```diff
@@ -3058,13 +3058,21 @@ std::shared_ptr<CWallet> CWallet::Create(WalletContext& context, const std::stri
             warnings.push_back(strprintf(_("%s is set very high! Fees this large could be paid on a single transaction."), "-maxtxfee"));
         }
 
-        if (chain && CFeeRate{max_fee.value(), 1000} < chain->relayMinFee()) {
```
I don't think we should remove this check for setting a `max_fee` that results in fee rates less than the `relayMinFee`.

ismaelsadeeq replied:
`relayMinFee` is relative to the fee rate of the transaction. I believe I did the right thing by removing it and enforcing this check on `-maxfeerate` instead.

I don't think we should use the base fee value as a quantifier for the fee rate, as transaction sizes differ. 
This is indicated in the issue description #29220.

ismaelsadeeq replied:
> Removing the check entirely for one of the options feels like a regression.

I don't think its a regression because this is the right behavior to check if the maximum fee rate we set is < minimum relay fee rate.


>  but previously this code was checking that the user did not set a `maxtxfee` that was so low that the minrelay fee could never be met. Now you've removed that check, which means I can set a very low maxtxfee and not get a warning.

With the assumption of 1k vb size, transactions can have more or less virtual size though.

I think we prevent creating transactions with fee rate below `relayMinFee` , so I think this okay?

glozow replied:
> Now you've removed that check, which means I can set a very low maxtxfee and not get a warning.

Perhaps just retain a version of the old check by enforcing that `maxtxfee` is set to at least `CAmount FLOOR_MAX_TXFEE{1000}` (a conversion from the default min relay feerate)?

josibake replied:
Summarizing my thoughts from some offline discussion with @ismaelsadeeq :

* I'm fairly certain we don't check `relayMinFee` when creating a transaction, but it would be good to confirm this (even better, to have a functional test!)
* If the user can set both `relayMinFee` and `maxtxfee`, I think we need to have a sanity check at startup to make sure they are not setting insane values that could result in them creating stuck transactions
* I think its okay to use a "fake transaction size" for a sanity check, but we should probably call it that i.e. `BASELINE_TX_SIZE = <1000, or maybe a minimum tx size?>`. That way, it is used consistently throughout the codebase, has a comment explaining what it used for and what it shouldn't be used for, etc
* There are a few other places where we use `maxtxfee` to calculate a fake feerate (such as `settxfee`). We should change those to use `maxtxfeerate`

In general, I agree we should always use `maxtxfeerate` when checking things about feerates. The only exception is during startup where we need to compare `maxtxfee` to other user set values relating to feerates.

josibake replied:
> Perhaps just retain a version of the old check by enforcing that maxtxfee is set to at least CAmount FLOOR_MAXTXFEE{1000} (a conversion from the default min relay feerate)?

This doesn't cover the case where a user sets a `maxtxfee` and a custom `relayMinFee`. Admittedly, the old check wasn't perfect in that the user could still create very large transactions that *might* still cause a problem. I think the best would be have a sanity check at startup (either the old one, or a new one with a different default for baseline tx size) and ALSO check that we are respecting the `relayMinFee` during transaction construction

ismaelsadeeq replied:
>I'm fairly certain we don't check relayMinFee when creating a transaction, but it would be good to confirm this (even better, to have a functional test!)

I've tested this, and it turns out that the wallet indeed prevents creating transactions below the `minrelaytxfee`
 https://github.com/bitcoin/bitcoin/blob/a11585692e72cac468fb1496ea2c30e4c07f73e5/src/wallet/spend.cpp#L1055
 
Here is a test that confirms we can't create transactions with a fee rate below the minimum relay fee, and also the node won't start if maxtxfee is very low so that a 1000 VB transaction with `maxtxfee` base fee fee rate is below`minrelaytxfee`.
See https://github.com/ismaelsadeeq/bitcoin/commit/aa075b8ef0b43f03d2fbef41c04a71403e93d904

In cases where users set a high `minrelaytxfee` and low `maxtxfee`, they can't even create transactions.
I agree with @josibake suggestion and added a commit that performs the same check in wallet initialization as before.
node won't start if `maxtxfee` is very low so that a 1000 vb transaction with `maxtxfee` base fee fee rate is below the `minrelaytxfee`.

In cases where the transaction size exceeds 1000 vb, the wallet check should safeguard against creating stuck transactions.

josibake reviewed (commented):
It seems odd to me that we would ever have `maxtxfee` and `maxfeerate` set together as they seem to cover entirely different scenarios.

For `maxtxfee`, a user is expressing "I don't or care what my transaction size is but I know I never want to spend more than X in total fees."

For `maxfeerate` as user is expressing "I know that my transactions will vary a lot in size, so I don't know what the total fees will be, but I do know that I don't want to pay more than X sats/per vbyte."

I am struggling to think of a scenario where it makes sense to combine them. On the other hand, if I set a `maxtxfee` of Y and a `maxfeerate` of X, there input sets that are valid for Y but not valid for X and also inputs that are valid for X but not for Y.

What do you think about only letting the user set one or the other? Or is there a use case for caring about the total fees you pay AND caring about the sats per vbyte you pay that I'm not understanding?

achow101 commented:
> Or is there a use case for caring about the total fees you pay AND caring about the sats per vbyte you pay that I'm not understanding?

I think it makes sense to have both. I could have a `maxtxfee` of the upper bound of what I'm willing to pay ever, and also set `maxfeerate` to something reasonable just to make sure I don't typo a feerate and accidentally pay a feerate orders of magnitude higher than intended. At `maxfeerate`, a small transaction can still pay a smaller fee than `maxtxfee`, but a large transaction could exceed `maxtxfee`.

josibake commented:
> I think it makes sense to have both. I could have a `maxtxfee` of the upper bound of what I'm willing to pay ever, and also set `maxfeerate` to something reasonable just to make sure I don't typo a feerate and accidentally pay a feerate orders of magnitude higher than intended. At `maxfeerate`, a small transaction can still pay a smaller fee than `maxtxfee`, but a large transaction could exceed `maxtxfee`.

But in this example, if you typo'd a feerate orders of magnitude higher, your `maxtxfee` should get hit. If you accidentally typo a higher fee rate than intended but it's below the `maxtxfee` you're willing to pay, that doesn't seem like a big problem to me.

Seems less footgunny to only let the user set one or the other (and then as a belt and suspenders still check the defaults for both, regardless of what is set?). But I might be overthinking this.. just seems really gross to have to now consider both of these in coin selection, especially if a user sets wonky values.

ismaelsadeeq commented:
Thank you for your comments @josibake @glozow @achow101 , will address review comments shortly.

> if you typo'd a feerate orders of magnitude higher, your maxtxfee should get hit.

I don't think this is true; it depends on what your `maxtxfee` is. Going by the default value of 0.10 BTC.

This is not likely to occur. 

I did a manual test on master.

When I set `maxtxfee` to 0.10 BTC on master, if you mistakenly type 10,000 as the fee rate, the max fee exceed error will not be hit.

However, in some cases where you set a low `maxtxfee`it may be hit or not depending on the transaction size.
It's uncertain.

```terminal
abubakarismail@Abubakars-MacBook-Pro bitcoin % ./src/bitcoind -regtest -maxtxfee=0.10  -daemon                                                                                          
Bitcoin Core starting

abubakarismail@Abubakars-MacBook-Pro bitcoin % ./src/bitcoin-cli -regtest loadwallet abubakar-test                                                                          
{
  "name": "abubakar-test"
}

abubakarismail@Abubakars-MacBook-Pro bitcoin % ./src/bitcoin-cli -regtest -named sendtoaddress address="bcrt1qt3fvqww6reduq287nhl59cu05hmyrj7achar44" fee_rate=10000 amount=1
22a92fdfa68581ba1d3da4aeedaeb20c7570b8c8eff9c29699c846aa1dacd73d

```


---

On this branch, if I set my `-maxtxfee` as 0.10 absolute fee and `-maxfeerate` as (0.010 which is 1000s/vb), and make a typo in a `sendtoaddress` call with a `fee_rate` of 10,000 s/vb instead of 100, the `maxfeerate` check will be hit.
```
 ./src/bitcoind -regtest -maxfeerate=0.01 -maxtxfee=0.10 -daemon                                                                         
Bitcoin Core starting
abubakarismail@Abubakars-MacBook-Pro bitcoin % ./src/bitcoin-cli -regtest loadwallet abubakar-test                                                                         
{
  "name": "abubakar-test"
}

abubakarismail@Abubakars-MacBook-Pro bitcoin % ./src/bitcoin-cli -regtest -named sendtoaddress address="bcrt1qt3fvqww6reduq287nhl59cu05hmyrj7achar44" fee_rate=10000 amount=1
error code: -6
error message:
Fee exceeds maximum configured by user (e.g. -maxtxfee, maxfeerate)
```


IIUC `-maxtxfee` should generally be high as the maximum absolute fee you think you can ever pay, and `-maxfeerate` as a cautionary check against an absurd fee rate for a transaction based on current fee estimates.

Maybe I should differentiate between the two error messages and make it more verbose, so that users can know which one is hit and adjust if its not intentional.

ismaelsadeeq commented:
> just seems really gross to have to now consider both of these in coin selection, especially if a user sets wonky values.

I dont really understand how this can affect coinselection, can you expand please? Thanks

josibake commented:
@ismaelsadeeq thanks for walking through the scenario! Reading over what you posted, if `maxtxfee` is left as the default, then the user can set whatever they want for `maxfeerate` without any issues. Depending on the types of transactions they make, they are essentially setting a new, lower `maxtxfee`.

If the user is setting both, they could create weird scenarios where the `maxtxfee` is too low compared to the `maxtxfeerate` and vice versa. But thinking it through and with your examples, I think most of my concern boils down to "the user set a `maxtxfee` which is too low for the current fee environment." 

> Maybe I should differentiate between the two error messages and make it more verbose, so that users can know which one is hit and adjust if its not intentional.

I think differentiating errors makes sense! Then it doesn't really matter if the user sets bad values for `max_fee_rate` or `max_tx_fee`, so long as we tell them which bad value is getting hit and prompt them to change it.

josibake commented on src/wallet/spend.cpp:
```diff
@@ -1294,6 +1294,11 @@ static util::Result<CreatedTransactionResult> CreateTransactionInternal(
 
+    CFeeRate tx_fee_rate = CFeeRate(current_fee, nBytes);
+    if (tx_fee_rate > wallet.m_max_tx_fee_rate) {
+        return util::Error{TransactionErrorString(TransactionError::MAX_FEE_EXCEEDED)};
```
perhaps introduce a new error type: 

```suggestion
        return util::Error{TransactionErrorString(TransactionError::MAX_FEE_RATE_EXCEEDED)};
```

ismaelsadeeq replied:
Added

murchandamus replied:
I realized now that you are adding the new Error type in the third PR. It would make more sense to me if the new Error type were introduced first, and then the test that uses it would get added rather than adding the test and then later introducing the new Error type.

ismaelsadeeq replied:
Thanks @murchandamus, I agree with your suggestion and reorder the commits.

josibake commented on src/util/error.cpp:
```diff
@@ -32,7 +32,9 @@ bilingual_str TransactionErrorString(const TransactionError err)
             return Untranslated("Specified sighash value does not match value stored in PSBT");
         case TransactionError::MAX_FEE_EXCEEDED:
-            return Untranslated("Fee exceeds maximum configured by user (e.g. -maxtxfee, maxfeerate)");
+            return Untranslated("Fee exceeds maximum configured by user.");
```
Better to keep the option name in the error message:

```suggestion
            return Untranslated("Fee exceeds maximum configured by user (maxtxfee)");
```

ismaelsadeeq replied:
Yes but I think should come in a follow-up after cleaning up your suggestion here
https://github.com/bitcoin/bitcoin/pull/29278#discussion_r1472790897

I have a local branch that does this on top of this.

ismaelsadeeq replied:
The reason is that the maximum fee rate check for [`sendrawtransaction`](https://github.com/bitcoin/bitcoin/blob/a11585692e72cac468fb1496ea2c30e4c07f73e5/src/rpc/mempool.cpp#L84-L94) is performed in `BroadcastTransaction` which returned [`MAX_FEE_EXCEEDED`](https://github.com/bitcoin/bitcoin/blob/a11585692e72cac468fb1496ea2c30e4c07f73e5/src/node/transaction.cpp#L71-L79) error type to `sendrawtransaction`.
If I change the `MAX_FEE_EXCEEDED` error message to "Fee exceeds maximum configured by user (maxtxfee)" without changing how it's checked, this would be misleading for `sendrawtransaction` callers and I wanted this to be a minimal wallet change that adds the `maxfeerate` option to the wallet.

However I agree with your comment, I have now added a commit that also checks the fee rate limit in `BroadcastTransaction`,  and changed how `sendrawtransaction` is implemented, and updated the `MAX_FEE_EXCEEDED` transaction error type message.

murchandamus commented:
Concept ACK

My scenario for using both `maxtxfee` and `maxfeerate` would be slightly different. For example looking at the mempool in January, I might generally be okay with paying up to 50 sat/vB, but not a higher feerate, because I expect that it will generally suffice to get a confirmation within 24h, but at such a high feerate, I would not want to make a transaction with more than 10 inputs. So, I’d delimit the feerate itself on the one hand, and then also limit the `maxtxfee` to 35'000 sats which is enough for a transaction with 10 P2TR keypath inputs and two P2TR outputs at 50 sat/vB. Through this general limit on the cost of a single transaction I could essentially set an upper bound on the total input weight of my transactions which would increase with lower feerates.

murchandamus commented on test/functional/rpc_psbt.py:
```diff
@@ -321,7 +321,7 @@ def run_test(self):
         self.log.info("Test invalid fee rate settings")
         for param, value in {("fee_rate", 100000), ("feeRate", 1)}:
-            assert_raises_rpc_error(-4, "Fee exceeds maximum configured by user (e.g. -maxtxfee, maxfeerate)",
+            assert_raises_rpc_error(-4, "Fee exceeds maximum configured by user.",
```
In d4dceeafc2292cd050cb10e308d73bc8ef84201a (wallet: test: add new transaction error type):

This seems strictly less informative. I would second @josibake in the suggestion to keep mentioning the `maxtxfee` here. Maybe I’m missing something, could you elaborate why you would push that back to a follow-up?

ismaelsadeeq replied:
This is fixed, resolving.

murchandamus reviewed (commented):
Concept ACK

ismaelsadeeq committed:
[doc]: add release notes

ismaelsadeeq commented on src/wallet/rpc/spend.cpp:
```diff
@@ -430,15 +430,14 @@ RPCHelpMan settxfee()
-    } else if (tx_fee_rate > max_tx_fee_rate) {
-        throw JSONRPCError(RPC_INVALID_PARAMETER, strprintf("txfee cannot be more than wallet max tx fee (%s)", max_tx_fee_rate.ToString()));
+    } else if (tx_fee_rate > pwallet->m_max_tx_fee_rate) {
+        throw JSONRPCError(RPC_INVALID_PARAMETER, strprintf("txfee cannot be more than wallet max tx fee rate (%s)", pwallet->m_max_tx_fee_rate.ToString()));
```
Fixed, thanks

ismaelsadeeq commented on src/wallet/wallet.cpp:
```diff
@@ -3058,13 +3058,21 @@ std::shared_ptr<CWallet> CWallet::Create(WalletContext& context, const std::stri
         }
 
-        if (chain && CFeeRate{max_fee.value(), 1000} < chain->relayMinFee()) {
+        walletInstance->m_default_max_tx_fee = max_fee.value();
```
fixed

ismaelsadeeq commented on test/functional/wallet_send.py:
```diff
@@ -184,6 +184,33 @@ def test_send(self, from_wallet, to_wallet=None, amount=None, data=None,
+        # All transactions with a fee rate below 10,000 will be created.
+
+        # All transaction with feerate <= 10,000 sats/vb can be created.
+        self.nodes[0].sendtoaddress(self.nodes[0].getnewaddress(), amount=1, fee_rate=1000)
```
I've implemented your suggestion 👍🏾

ismaelsadeeq commented on test/functional/wallet_send.py:
```diff
@@ -184,6 +184,33 @@ def test_send(self, from_wallet, to_wallet=None, amount=None, data=None,
+
+        # Fee rates <= 10 sats/vb will be accepted by the wallet.
+        self.nodes[0].sendtoaddress(self.nodes[0].getnewaddress(), amount=1, fee_rate=5)
+        self.nodes[0].sendtoaddress(self.nodes[0].getnewaddress(), amount=1, fee_rate=6)
```
Fixed

ismaelsadeeq commented on src/dummywallet.cpp:
```diff
@@ -37,6 +37,7 @@ void DummyWalletInit::AddWalletOptions(ArgsManager& argsman) const
         "-keypool=<n>",
         "-maxapsfee=<n>",
         "-maxtxfee=<amt>",
+        "-maxfeerate=<amt>",
```
Added thanks

ismaelsadeeq commented on src/wallet/wallet.h:
```diff
@@ -136,6 +136,8 @@ static const bool DEFAULT_DISABLE_WALLET = false;
 //! -maxtxfee default
 constexpr CAmount DEFAULT_TRANSACTION_MAXFEE{COIN / 10};
+//! maxfeerate default
+const CFeeRate DEFAULT_TRANSACTION_MAXFEERATE(CFeeRate(COIN / 10));
```
Yes its 10'000 sat/vb, I agree its high indeed Should I reduce it to maybe 1000 s/vb instead?

---
This is configurable option can be reduced by users to threshold they want.

murchandamus replied:
Yeah, I think that currently a `1000 sat/vB` limit would be better. Also see #29661.

ismaelsadeeq replied:
> but that might be insufficient for a child transaction bumping a much larger parent. 

Users can then restart with a much higher `maxfeerate` whenever they want to feebump a transaction and the limit is hit? but  I reckon restarting may be undesirable.

---
Anyways whether it's `1000` or `2000` `sat/vB` a sane  maximum transaction fee rate is a function of Block space demand which is unpredictable, what is insane fee rate maybe sane.   Thats why users can pass a lower limit or higher limit based on that factor?

Should I leave as is and continue discussion about this in https://github.com/bitcoin/bitcoin/issues/29661?

ismaelsadeeq commented on src/wallet/feebumper.cpp:
```diff
@@ -80,6 +80,13 @@ static feebumper::Result CheckFeeRate(const CWallet& wallet, const CMutableTrans
         return feebumper::Result::WALLET_ERROR;
     }
 
+    // check that new fee rate does not exceed maxfeerate
```
> compare it with the resulting mining score instead of the new transaction’s individual feerate?

🤔 Not sure how to implement the change you suggested honestly.
The current behavour is checking if target fee rate of user wants to bump the transaction to is greater than `maxfeerate` which I believe is the new mining score of the bumped transaction no? 
I might be wrong on my assumption though, would you clarify your comment please? thanks.

murchandamus replied:
Yeah, it will probably be easier to introduce that in the context of a _cluster mempool_ where we would much more easily be able to get a read on the effective feerate at which a transaction would be up for block-inclusion.

josibake reviewed (commented):
Looking good, and thanks for adding more tests!

ismaelsadeeq commented on src/wallet/wallet.cpp:
```diff
@@ -1982,7 +1982,7 @@ bool CWallet::SubmitTxMemoryPoolAndRelay(CWalletTx& wtx, std::string& err_string
     // If transaction was previously in the mempool, it should be updated when
     // TransactionRemovedFromMempool fires.
-    bool ret = chain().broadcastTransaction(wtx.tx, m_default_max_tx_fee, /*max_tx_feerate*/ CFeeRate(0), relay, err_string);
+    bool ret = chain().broadcastTransaction(wtx.tx, m_default_max_tx_fee, m_max_tx_fee_rate, relay, err_string);
```
Yes 👍🏾  its better that way.

ismaelsadeeq commented on test/functional/wallet_bumpfee.py:
```diff
@@ -545,9 +545,9 @@ def test_settxfee(self, rbf_node, dest_address):
     assert_greater_than(Decimal("0.00001000"), abs(requested_feerate - actual_feerate))
     rbf_node.settxfee(Decimal("0.00000000"))  # unset paytxfee
 
-    # check that settxfee respects -maxtxfee
-    self.restart_node(1, ['-maxtxfee=0.000025'] + self.extra_args[1])
-    assert_raises_rpc_error(-8, "txfee cannot be more than wallet max tx fee", rbf_node.settxfee, Decimal('0.00003'))
```
in "wallet: add maxfeerate wallet startup option" https://github.com/bitcoin/bitcoin/commit/4165dc41289fd337e279f3986dfc1a84345c900e

After adding `maxfeerate` option we now only check if the fee rate user set is less  `minrelaytxfee`, less than wallet min fee, or whether it exceed `maxfeerate`.
So their is no need to check for `maxtxfee` here, because its not enforced here.

Will add this info in commit message, thanks

ismaelsadeeq replied:
Added a test for this thanks.

ismaelsadeeq commented on src/wallet/rpc/spend.cpp:
```diff
@@ -1485,6 +1485,9 @@ RPCHelpMan sendall()
                 throw JSONRPCError(RPC_WALLET_ERROR, TransactionErrorString(TransactionError::MAX_FEE_EXCEEDED).original);
             }
+            if (CFeeRate(fee_from_size, tx_size.vsize) > pwallet->m_max_tx_fee_rate) {
+                throw JSONRPCError(RPC_WALLET_ERROR, TransactionErrorString(TransactionError::MAX_FEE_EXCEEDED).original);
```
Yes 👍🏾

ismaelsadeeq replied:
Fixed

ismaelsadeeq commented on test/functional/wallet_bumpfee.py:
```diff
@@ -130,7 +130,7 @@ def test_invalid_parameters(self, rbf_node, peer_node, dest_address):
 
         self.log.info("Test invalid fee rate settings")
-        assert_raises_rpc_error(-4, "Specified or calculated fee 0.141 is too high (cannot be higher than -maxtxfee 0.10",
+        assert_raises_rpc_error(-4, "New fee rate 1.00 is too high (cannot be higher than -maxfeerate 0.10)",
```
Because after https://github.com/bitcoin/bitcoin/commit/eadc3a4500d6b7be263a2981602c20f6e5eb13a4 we now check for `maxfeerate` threshold first before checking `maxtxfee` threshold.
And the specified fee rate is above default `maxferate`, as such it will be hit first.

But I agree, I will add another test where  specified bump fee rate is below the `maxfeerate` but we hit `maxtxfee` due to size or maybe just restart with `maxtxfee`  set to a lower value ?

josibake replied:
Either sounds fine! The main point is we shouldn't be removing test cases for maxtxfee while adding maxtxfeerate.

ismaelsadeeq replied:
Thanks updated the tests.

ismaelsadeeq commented on src/rpc/mempool.cpp:
```diff
@@ -85,13 +85,11 @@ static RPCHelpMan sendrawtransaction()
             AssertLockNotHeld(cs_main);
             NodeContext& node = EnsureAnyNodeContext(request.context);
-            const TransactionError err = BroadcastTransaction(node, tx, err_string, max_raw_tx_fee, /*relay=*/true, /*wait_callback=*/true);
+            const TransactionError err = BroadcastTransaction(node, tx, err_string, /*max_raw_fee*/ 0, max_raw_tx_fee_rate, /*relay=*/true, /*wait_callback=*/true);
```
Fixed

ismaelsadeeq commented on src/wallet/test/wallet_tests.cpp:
```diff
@@ -820,7 +821,7 @@ BOOST_FIXTURE_TEST_CASE(CreateWallet, TestChain100Setup)
     m_coinbase_txns.push_back(CreateAndProcessBlock({block_tx}, GetScriptForRawPubKey(coinbaseKey.GetPubKey())).vtx[0]);
     auto mempool_tx = TestSimpleSpend(*m_coinbase_txns[1], 0, coinbaseKey, GetScriptForRawPubKey(key.GetPubKey()));
-    BOOST_CHECK(m_node.chain->broadcastTransaction(MakeTransactionRef(mempool_tx), DEFAULT_TRANSACTION_MAXFEE, false, error));
+    BOOST_CHECK(m_node.chain->broadcastTransaction(MakeTransactionRef(mempool_tx), DEFAULT_TRANSACTION_MAXFEE, /*max_tx_feerate*/ CFeeRate(0), false, error));
```
Updated 👍🏾

ismaelsadeeq commented on src/qt/walletmodel.cpp:
```diff
@@ -231,7 +231,7 @@ WalletModel::SendCoinsReturn WalletModel::prepareTransaction(WalletModelTransact
 
         // Reject absurdly high fee. (This can never happen because the
         // wallet never creates transactions with fee greater than
-        // m_default_max_tx_fee. This merely a belt-and-suspenders check).
```
Done

ismaelsadeeq commented on src/wallet/wallet.cpp:
```diff
@@ -3056,6 +3056,16 @@ std::shared_ptr<CWallet> CWallet::Create(WalletContext& context, const std::stri
+        // Otherwise transactions with fee rate greater than or equal to the minrelaytxfee will likely exceed maxtxfee.
+        // In such cases, the wallet won't be able to create transactions. Therefore, shut down early.
+        } else if (chain && CFeeRate(max_fee.value(), 1000) < chain->relayMinFee()) {
+            error = strprintf(_("Invalid amount for %s=<amount>: '%s' is too low. It won't allow creating transactions exceeding the minrelay fee of %s. Consider adjusting %s or %s."),
```
Fixed

ismaelsadeeq commented:
Thanks for your review @josibake 
I fixed all review comments
Forced pushed from a5d10367cf832497af2ac72f8c2c42dd25398c63 to 1fae882f98205afb05e96432ce57212226bc0909 [Compare changes](https://github.com/bitcoin/bitcoin/compare/a5d10367cf832497af2ac72f8c2c42dd25398c63..1fae882f98205afb05e96432ce57212226bc0909)

josibake commented on src/rpc/mempool.cpp:
```diff
@@ -85,13 +85,11 @@ static RPCHelpMan sendrawtransaction()
             AssertLockNotHeld(cs_main);
             NodeContext& node = EnsureAnyNodeContext(request.context);
-            const TransactionError err = BroadcastTransaction(node, tx, err_string, max_raw_tx_fee, /*relay=*/true, /*wait_callback=*/true);
+            const TransactionError err = BroadcastTransaction(node, tx, err_string, /*max_tx_fee*/0, max_raw_tx_fee_rate, /*relay=*/true, /*wait_callback=*/true);
```
in "[node]: update BroadcastTransaction to check fee rate limit" (https://github.com/bitcoin/bitcoin/commit/eedbeccbdb274841fc2cf2c53931e58fbea50dfc):

nit:
```suggestion
            const TransactionError err = BroadcastTransaction(node, tx, err_string, /*max_tx_fee=*/0, max_raw_tx_fee_rate, /*relay=*/true, /*wait_callback=*/true);
```

ismaelsadeeq replied:
Fixed

josibake reviewed (commented):
Looking good! I think moving https://github.com/bitcoin/bitcoin/pull/29278/commits/d83ecdde0dfdbf46806be1310ad63716e33edc96 to be the 3rd commit (right after introducing the new option) makes more sense as there are still at least a few spots where you return a `MAX_TX_FEE` error incorrectly, and then fix it up in a later commit.

Also, small typo in https://github.com/bitcoin/bitcoin/pull/29278/commits/eedbeccbdb274841fc2cf2c53931e58fbea50dfc:

s/differenciating/differentiating/

ismaelsadeeq commented on src/node/transaction.h:
```diff
@@ -39,11 +39,12 @@ static const CFeeRate DEFAULT_MAX_RAW_TX_FEE_RATE{COIN / 10};
  * @param[in]  tx the transaction to broadcast
  * @param[out] err_string reference to std::string to fill with error string if available
  * @param[in]  max_tx_fee reject txs with fees higher than this (if 0, accept any fee)
+ * @param[in]  max_tx_feerate reject txs with fee rate higher than this (if CFeeRate(0), accept any fee)
```
Thanks updated

ismaelsadeeq commented on src/node/transaction.cpp:
```diff
@@ -68,13 +69,15 @@ TransactionError BroadcastTransaction(NodeContext& node, const CTransactionRef t
+                } else if (max_tx_fee > 0 && result.m_base_fees.value() > max_tx_fee) {
+                    return TransactionError::MAX_FEE_EXCEEDED;
+                } else if (max_tx_feerate > CFeeRate(0) && CFeeRate(result.m_base_fees.value(), result.m_vsize.value()) > max_tx_feerate) {
                     return TransactionError::MAX_FEE_EXCEEDED;
```
The `MAX_FEERATE_EXCEEDED` error type was not introduced yet at this point.
 This is correct because at https://github.com/bitcoin/bitcoin/commit/eedbeccbdb274841fc2cf2c53931e58fbea50dfc, the error message indicates that the failure might be from `-maxtxfee` or `maxfeerate`.

And this is fixed in next commit 0ad2f2c32460fef9b8cd59e3920fcfd6602b02a4 after we introduced the `MAX_FEERATE_EXCEEDED` error type.

The options are;
- Introduce the error type first, then update` BroadcastTransaction` to support fee rate checks. This approach however, would result in misleading error messages in `sendrawtransaction`.
- Combine both changes in a single commit. 
- The current approach, update `BroadcastTransaction` to support checking fee rate then introduce the new error type which in my opinion, is the better approach.

josibake replied:
> The current approach, update BroadcastTransaction to support checking fee rate then introduce the new error type which in my opinion, is the better approach.

sounds good!

josibake commented:
CI failure doesn't make any sense given the lines that you changed in your latest force push. Could be a transient failure or a silent merge conflict? FWIW, I rebased locally on master and ran the functional tests without any failures.

josibake commented on src/qt/walletmodel.cpp:
```diff
@@ -231,7 +231,7 @@ WalletModel::SendCoinsReturn WalletModel::prepareTransaction(WalletModelTransact
 
         // Reject absurdly high fee. (This can never happen because the
         // wallet never creates transactions with fee greater than
-        // m_default_max_tx_fee. This merely a belt-and-suspenders check).
```
in "[refactor]: rename m_default_max_tx_fee to m_max_tx_fee" (https://github.com/bitcoin/bitcoin/pull/29278/commits/f1ff2b8f40edffc2b61240a976cff9a9fd51a110):

nit: this commit could be a scripted-diff

ismaelsadeeq replied:
Thanks, changed to a scripted-diff

josibake commented on src/wallet/rpc/spend.cpp:
```diff
@@ -439,8 +439,8 @@ RPCHelpMan settxfee()
         throw JSONRPCError(RPC_INVALID_PARAMETER, strprintf("txfee cannot be less than wallet min fee (%s)", pwallet->m_min_fee.ToString()));
-    } else if (tx_fee_rate > max_tx_fee_rate) {
-        throw JSONRPCError(RPC_INVALID_PARAMETER, strprintf("txfee cannot be more than wallet max tx fee (%s)", max_tx_fee_rate.ToString()));
+    } else if (tx_fee_rate > pwallet->m_max_tx_fee_rate) {
```
in "[wallet]: add maxfeerate wallet startup option" (https://github.com/bitcoin/bitcoin/pull/29278/commits/6a2c7722698d22c95220daf7727f9219ed843042):

The old `max_tx_fee_rate` on L435 should be deleted: 

https://github.com/bitcoin/bitcoin/pull/29278/commits/6a2c7722698d22c95220daf7727f9219ed843042#diff-26141d9c7da21eeb4b9e3ffedfaad83212d4710a9e62888f7abea076ca1d0538R435

ismaelsadeeq replied:
Deleted 👍🏾

josibake commented on src/interfaces/chain.h:
```diff
@@ -212,6 +212,7 @@ class Chain
     //! Return false if the transaction could not be added due to the fee or for another reason.
     virtual bool broadcastTransaction(const CTransactionRef& tx,
         const CAmount& max_tx_fee,
+        const CFeeRate& max_tx_feerate,
```
in "[node]: update BroadcastTransaction to check fee rate limit" (https://github.com/bitcoin/bitcoin/pull/29278/commits/368d140db549387fb7bf162c96dd9e9064045103):

nit: in other places, you've been calling this `max_tx_fee_rate`.

ismaelsadeeq replied:
Fixed

josibake commented on doc/release-notes-29278.md:
```diff
@@ -0,0 +1,7 @@
+========================
+
+- A new wallet startip option `-maxfeerate` is added.
+- This option sets the upper limit for wallet transaction fee rate.
+- The wallet will now refrain from creating transactions with a feerate exceeding the   `maxfeerate`.
```
in "[doc]: add release notes" (https://github.com/bitcoin/bitcoin/pull/29278/commits/bf6fc126d1e0b37e490d38124945fa4e8f7475f4):

nit: inconsistent usage of "fee rate" vs "feerate." considering "fee rate" is what you use in the error messages, that should probably be the one we stick with?

ismaelsadeeq replied:
Fixed the inconsistencies added in this PR, I am sticking to fee rate as u suggested, there are still others remaining in the codebase fixing that it is beyond the scope of this PR.

josibake reviewed (commented):
Overall, looking good! Left some nits and I noticed inconsistent usage of `feerate` vs `fee_rate` and "feerate" vs "fee rate" in the PR. I'm not sure which is more correct, considering I've seen both used in the wild, but "fee rate" / `fee_rate` _feels_ more correct?

ismaelsadeeq committed:
scripted-diff: rename `m_default_max_tx_fee` to `m_max_tx_fee`

-BEGIN VERIFY SCRIPT-
git grep -l "m_default_max_tx_fee" src | xargs sed -i "s/m_default_max_tx_fee/m_max_tx_fee/g"
git grep -l "getDefaultMaxTxFee" src | xargs sed -i "s/getDefaultMaxTxFee/getMaxTxFee/g"
-END VERIFY SCRIPT-

- The value is not always the default but can be configured
  during startup so `m_max_tx_fee` is the right name not
  `m_default_max_tx_fee`.

josibake commented:
ACK https://github.com/bitcoin/bitcoin/pull/29278/commits/879c048543b1f0239b4650dd0eea2f5215877916

ismaelsadeeq commented:
Rebased to fix merge conflict

murchandamus commented on src/wallet/wallet.cpp:
```diff
@@ -3118,6 +3118,16 @@ std::shared_ptr<CWallet> CWallet::Create(WalletContext& context, const std::stri
+        // Otherwise transactions with fee rate greater than or equal to the minrelaytxfee will likely exceed maxtxfee.
+        // In such cases, the wallet won't be able to create transactions. Therefore, shut down early.
+        } else if (chain && CFeeRate(max_fee.value(), 1000) < chain->relayMinFee()) {
+            error = strprintf(_("Invalid amount for %s=<amount>: '%s' conflicts with the minrelay fee of %s. Consider adjusting %s or %s."),
```
In "[wallet]: ensure 1kvb tx with -maxtxfee base fee has fee rate atleast `minrelaytxfee`" (e5b1b43bfdb82fc1b5bc19af04ede21d0de0ff38):

1. "minrelay fee" is a feerate, not a fee: How about: "[…] conflicts with the minimum relay transaction feerate […]"
2. "Consider" is a bit funny given that the `max_fee` amount was rejected, how about: "Please set a higher `maxtxfee` or lower `-minrelaytxfee` before retrying."

ismaelsadeeq replied:
Reworded to your suggestion

ismaelsadeeq committed:
[wallet]: update `max_fee` to `max_tx_fee`

- Also update `-maxapsfee` option from `max_fee` to `max_aps_fee`.
  This fixes the ambiguity in the variable name.

ismaelsadeeq commented on src/qt/walletmodel.cpp:
```diff
@@ -227,7 +227,7 @@ WalletModel::SendCoinsReturn WalletModel::prepareTransaction(WalletModelTransact
 
```
Fixed

ismaelsadeeq commented on src/wallet/wallet.cpp:
```diff
@@ -3120,13 +3120,22 @@ std::shared_ptr<CWallet> CWallet::Create(WalletContext& context, const std::stri
         }
 
-        if (chain && CFeeRate{max_fee.value(), 1000} < chain->relayMinFee()) {
+        walletInstance->m_max_tx_fee = max_fee.value();
```
I added a commit that does that https://github.com/bitcoin/bitcoin/pull/29278/commits/ef8403b79585fd1dfd6e8b0a68923985268be16f

Thanks.

ismaelsadeeq commented on src/wallet/wallet.cpp:
```diff
@@ -3120,13 +3120,22 @@ std::shared_ptr<CWallet> CWallet::Create(WalletContext& context, const std::stri
         }
 
-        walletInstance->m_max_tx_fee = max_fee.value();
+        walletInstance->m_max_tx_fee_rate = CFeeRate(max_fee_rate.value());
```
Fixed

ismaelsadeeq commented on test/functional/wallet_send.py:
```diff
@@ -184,6 +184,30 @@ def test_send(self, from_wallet, to_wallet=None, amount=None, data=None,
+    def test_maxfeerate(self):
+        self.log.info("test -maxfeerate enforcement on wallet transactions.")
+        # Default maxfeerate is 10,000 sats/vb
+        # Wallet will reject all transactions with fee rate above 10,000 sats/vb.
```
I agree with your suggestion. Fixed

ismaelsadeeq commented on src/wallet/wallet.cpp:
```diff
@@ -3118,6 +3118,16 @@ std::shared_ptr<CWallet> CWallet::Create(WalletContext& context, const std::stri
+
+        // Wallet prevents creating transactions with fee rates lower than minrelaytxfee.
+        // Also the wallet prevents creating transaction with base fee above maxtxfee.
+        // Ensure that a 1kvb transaction, with a base fee set to maxtxfee, maintains a fee rate of at least minrelaytxfee.
```
Fixed, it's now a warning instead a startup failure..

ismaelsadeeq commented on src/wallet/feebumper.cpp:
```diff
@@ -80,6 +80,13 @@ static feebumper::Result CheckFeeRate(const CWallet& wallet, const CMutableTrans
+    // check that new fee rate does not exceed maxfeerate
+    if (newFeerate > wallet.m_max_tx_fee_rate) {
+        errors.push_back(strprintf(Untranslated("New fee rate %s is too high (cannot be higher than -maxfeerate %s)"),
+            FormatMoney(newFeerate.GetFeePerK()), FormatMoney(wallet.m_max_tx_fee_rate.GetFeePerK())));
```
Fixed

ismaelsadeeq commented on src/wallet/init.cpp:
```diff
@@ -66,6 +66,7 @@ void WalletInit::AddWalletOptions(ArgsManager& argsman) const
     argsman.AddArg("-maxapsfee=<n>", strprintf("Spend up to this amount in additional (absolute) fees (in %s) if it allows the use of partial spend avoidance (default: %s)", CURRENCY_UNIT, FormatMoney(DEFAULT_MAX_AVOIDPARTIALSPEND_FEE)), ArgsManager::ALLOW_ANY, OptionsCategory::WALLET);
     argsman.AddArg("-maxtxfee=<amt>", strprintf("Maximum total fees (in %s) to use in a single wallet transaction; setting this too low may abort large transactions (default: %s)",
         CURRENCY_UNIT, FormatMoney(DEFAULT_TRANSACTION_MAXFEE)), ArgsManager::ALLOW_ANY, OptionsCategory::DEBUG_TEST);
+    argsman.AddArg("-maxfeerate=<amt>", strprintf("Maximum fee rate (in %s) of a single wallet transaction (default: %s)", CURRENCY_UNIT, FormatMoney(DEFAULT_TRANSACTION_MAXFEERATE.GetFeePerK())), ArgsManager::ALLOW_ANY, OptionsCategory::WALLET);
```
Thats right reworded to"Maximum fee rate (in `BTC/kvB`) of a single wallet transaction (default: `0.10` `BTC/kvB`)"  thank you

ismaelsadeeq commented on src/wallet/wallet.h:
```diff
@@ -136,6 +136,8 @@ static const bool DEFAULT_DISABLE_WALLET = false;
 //! -maxtxfee default
 constexpr CAmount DEFAULT_TRANSACTION_MAXFEE{COIN / 10};
+//! maxfeerate default
+const CFeeRate DEFAULT_TRANSACTION_MAXFEERATE(CFeeRate(COIN / 10));
```
Fixed thank you

ismaelsadeeq commented on test/functional/wallet_bumpfee.py:
```diff
@@ -545,9 +545,9 @@ def test_settxfee(self, rbf_node, dest_address):
-    # check that settxfee respects -maxtxfee
-    self.restart_node(1, ['-maxtxfee=0.000025'] + self.extra_args[1])
-    assert_raises_rpc_error(-8, "txfee cannot be more than wallet max tx fee", rbf_node.settxfee, Decimal('0.00003'))
+    # check that settxfee respects -maxfeerate
```
I think should be updated to `settxfeerate` in a followup?

ismaelsadeeq commented:
Thanks for your review @josibake @murchandamus 

Force pushed from https://github.com/bitcoin/bitcoin/commit/879c048543b1f0239b4650dd0eea2f5215877916 to https://github.com/bitcoin/bitcoin/commit/50d2a010fcd57b58537b711846340c68f6614720 to address recent review comments by @murchandamus  [compare differences](https://github.com/bitcoin/bitcoin/compare/879c048543b1f0239b4650dd0eea2f5215877916..50d2a010fcd57b58537b711846340c68f6614720)

Changes 
- Rebased for green C.I
- Updated https://github.com/bitcoin/bitcoin/commit/e602fd275f3632fafa9416281cc375ed622e0876 to fix commit message typo
- Updated `max_fee` to `max_tx_fee` to fix ambiguity
- Updated `max_fee_rate` to `max_tx_feerate` to have consistent variable name
- Also Updated `DEFAULT_TRANSACTION_MAXFEERATE` to `DEFAULT_MAX_TRANSACTION_FEERATE`  for consistency in naming
- Fix incorrect `maxfeerate` option description to "Maximum fee rate (in BTC/kvB) of a single wallet transaction (default: 0.10 BTC/kvB)".
- Return more verbose error if fee bumping fail when `maxfeerate` limit is hit
- Updates `sats/vb` to a more accurate description `sat/vB`
- Fix https://github.com/bitcoin/bitcoin/pull/29278#discussion_r1526777732 comment by returning a warning when creating or loading wallet instead of failing at startup.

ismaelsadeeq commented:
Rebased to fix merge conflict.

ismaelsadeeq commented:
> https://cirrus-ci.com/task/4669206682140672?logs=ci#L3887

thanks @maflcko, will fix the C.I

ismaelsadeeq commented:
@DrahtBot  I think C.I failure here is unrelated? 
I was not successful in my attempt to recreate it locally.

ismaelsadeeq commented:
@maflcko Interestingly I got the error after 5 runs.
<details>

<summary>logs</summary>

```terminal
2024-11-15T19:20:59.167000Z TestFramework (INFO): Test issue 22670 ApproximateBestSubset bug
2024-11-15T19:21:00.082000Z TestFramework (ERROR): JSONRPC error
Traceback (most recent call last):
  File "/Users/abubakarismail/Desktop/Work/bitcoin-dev/bitcoin/test/functional/test_framework/test_framework.py", line 132, in main
    self.run_test()
  File "/Users/abubakarismail/Desktop/Work/bitcoin-dev/bitcoin/build/test/functional/wallet_fundrawtransaction.py", line 150, in run_test
    self.test_22670()
  File "/Users/abubakarismail/Desktop/Work/bitcoin-dev/bitcoin/build/test/functional/wallet_fundrawtransaction.py", line 1434, in test_22670
    do_fund_send(lower_bound)
  File "/Users/abubakarismail/Desktop/Work/bitcoin-dev/bitcoin/build/test/functional/wallet_fundrawtransaction.py", line 1419, in do_fund_send
    funded_tx = tester.fundrawtransaction(create_tx)
                ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/Users/abubakarismail/Desktop/Work/bitcoin-dev/bitcoin/test/functional/test_framework/coverage.py", line 50, in __call__
    return_val = self.auth_service_proxy_instance.__call__(*args, **kwargs)
                 ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/Users/abubakarismail/Desktop/Work/bitcoin-dev/bitcoin/test/functional/test_framework/authproxy.py", line 146, in __call__
    raise JSONRPCException(response['error'], status)
test_framework.authproxy.JSONRPCException: Insufficient funds (-4)
2024-11-15T19:21:00.144000Z TestFramework (INFO): Stopping nodes
2024-11-15T19:21:00.422000Z TestFramework (WARNING): Not cleaning up dir /var/folders/dj/d8p8jhd172n7wnq81ryfl6rc0000gn/T/bitcoin_func_test_per45s2i
2024-11-15T19:21:00.422000Z TestFramework (ERROR): Test failed. Test logging available at /var/folders/dj/d8p8jhd172n7wnq81ryfl6rc0000gn/T/bitcoin_func_test_per45s2i/test_framework.log
2024-11-15T19:21:00.422000Z TestFramework (ERROR): 
2024-11-15T19:21:00.422000Z TestFramework (ERROR): Hint: Call /Users/abubakarismail/Desktop/Work/bitcoin-dev/bitcoin/test/functional/combine_logs.py '/var/folders/dj/d8p8jhd172n7wnq81ryfl6rc0000gn/T/bitcoin_func_test_per45s2i' to consolidate all logs
2024-11-15T19:21:00.422000Z TestFramework (ERROR): 
2024-11-15T19:21:00.422000Z TestFramework (ERROR): If this failure happened unexpectedly or intermittently, please file a bug and provide a link or upload of the combined log.
2024-11-15T19:21:00.422000Z TestFramework (ERROR): https://github.com/bitcoin/bitcoin/issues
2024-11-15T19:21:00.422000Z TestFramework (ERROR): 
```

</details>

murchandamus reviewed (approved):
ACK e5d1d25912390954b1a064c5c1f9c5d2c1726430

I think the MAXFEE ↦ MINFEE test issue should still be amended, the rest of my comments are not urgent

ismaelsadeeq committed:
[wallet]: add `maxfeerate` wallet startup option

- The commits adds a new wallet startup option `-maxfeerate`

- The value will be used as the upper limit of wallet transaction fee rate.

- The commit updates all instances where `-maxtxfee` value is utilized to check
  wallet transactions fee rate upper limit to now use `-maxfeerate` value.

ismaelsadeeq committed:
[node]: update `BroadcastTransaction` to check fee rate limit

`BroadcastTransaction` now accepts a fee rate limit as an additional input
parameter. It rejects transactions whose fee exceeds the maximum fee amount
permitted by that fee rate for the transaction vsize.

Compare against `maxfeerate.GetFee(vsize)` instead of reconstructing a fee
rate from the accepted fee. This preserves the existing rounded fee cap
semantics used by `testmempoolaccept` and avoids rejecting transactions whose
rounded fee is still within the configured limit.

This allows callers to distinguish between failures caused by the absolute fee
limit (`-maxtxfee`) and the fee rate limit (`maxfeerate` or `-maxfeerate`).

`TestSimpleSpend` creates transactions paying `DEFAULT_TRANSACTION_MAXFEE`.
Since `BroadcastTransaction` now also checks the fee rate limit, these small
high-fee transactions would exceed `DEFAULT_MAX_TRANSACTION_FEERATE`. Update
the test broadcast limit to be slightly above the transaction fee rate.

In `wallet_fundrawtransaction.py`'s `test_22670` subtest, restart node 0 with
`-maxfeerate` above `-minrelaytxfee` so high-feerate transactions can be
broadcast.

ismaelsadeeq committed:
[util]: add a new transaction error type

- This distinguishes maxfeerate and maxtxfee error messages.

ismaelsadeeq committed:
[wallet]: warn when `-maxtxfee` conflicts with `-minrelaytxfee`

- Wallets cannot create transactions with fee rates below
  `-minrelaytxfee`, and they also reject transactions whose total fee
  exceeds `-maxtxfee`.

- Warn when a 1 kvB transaction paying exactly `-maxtxfee` would still
  have a fee rate below `-minrelaytxfee`. In this configuration, some
  transactions that meet the minimum relay fee rate may exceed
  `-maxtxfee`, preventing the wallet from creating them.

- Keep this warning independent from the existing high `-maxtxfee`
  warning, so users see both warnings when both conditions apply.

ismaelsadeeq commented on src/wallet/init.cpp:
```diff
@@ -64,6 +64,7 @@ void WalletInit::AddWalletOptions(ArgsManager& argsman) const
     argsman.AddArg("-maxapsfee=<n>", strprintf("Spend up to this amount in additional (absolute) fees (in %s) if it allows the use of partial spend avoidance (default: %s)", CURRENCY_UNIT, FormatMoney(DEFAULT_MAX_AVOIDPARTIALSPEND_FEE)), ArgsManager::ALLOW_ANY, OptionsCategory::WALLET);
     argsman.AddArg("-maxtxfee=<amt>", strprintf("Maximum total fees (in %s) to use in a single wallet transaction; setting this too low may abort large transactions (default: %s)",
         CURRENCY_UNIT, FormatMoney(DEFAULT_TRANSACTION_MAXFEE)), ArgsManager::ALLOW_ANY, OptionsCategory::DEBUG_TEST);
+    argsman.AddArg("-maxfeerate=<amt>", strprintf("Maximum fee rate (in %s/kvB) of a single wallet transaction (default: %s %s/kvB)", CURRENCY_UNIT, FormatMoney(DEFAULT_MAX_TRANSACTION_FEERATE.GetFeePerK()), CURRENCY_UNIT), ArgsManager::ALLOW_ANY, OptionsCategory::WALLET);
```
Taken!

ismaelsadeeq commented on test/functional/wallet_bumpfee.py:
```diff
@@ -571,6 +571,20 @@ def test_maxtxfee_fails(self, rbf_node, dest_address):
     # When user passed fee rate causes base fee to be above maxtxfee we fail early
     assert_raises_rpc_error(-4, "Specified or calculated fee 0.0000282 is too high (cannot be higher than -maxtxfee 0.000025)", rbf_node.bumpfee, rbfid, fee_rate=20)
+
+    self.log.info("Test that a low -maxtxfee that will likely prevents txs fee rate to reach -minrelaytxfee should warn users")
```
I attempt to rephrase the sentence!
Also happy to take explicit suggestion

ismaelsadeeq commented on src/wallet/test/wallet_tests.cpp:
```diff
@@ -49,7 +50,7 @@ BOOST_FIXTURE_TEST_SUITE(wallet_tests, WalletTestingSetup)
 {
     CMutableTransaction mtx;
-    mtx.vout.emplace_back(from.vout[index].nValue - DEFAULT_TRANSACTION_MAXFEE, pubkey);
+    mtx.vout.emplace_back(from.vout[index].nValue - DEFAULT_TRANSACTION_MINFEE, pubkey);
```
Done, 
I've switched this to`MAXFEERATE` and pass the tx fee rate as the max.

ismaelsadeeq commented on test/functional/rpc_rawtransaction.py:
```diff
@@ -400,7 +400,7 @@ def sendrawtransaction_tests(self):
     def sendrawtransaction_testmempoolaccept_tests(self):
         self.log.info("Test sendrawtransaction/testmempoolaccept with maxfeerate")
-        fee_exceeds_max = "Fee exceeds maximum configured by user (e.g. -maxtxfee, maxfeerate)"
+        fee_exceeds_max = "Fee rate exceeds maximum configured by user (maxfeerate)"
```
Done

ismaelsadeeq commented:
Thanks for review @murchandamus 

Forced pushed from e5d1d25912390954b1a064c5c1f9c5d2c1726430 to 46f0bc5b485130d3e7d14e6855d81d387950cb39 see [diff](https://github.com/bitcoin/bitcoin/compare/e5d1d25912390954b1a064c5c1f9c5d2c1726430..46f0bc5b485130d3e7d14e6855d81d387950cb39)

Changes

1. Improve the description of `maxfeerate` startup option
2. Update `sendrawtransaction_testmempoolaccept_tests`  `fee_exceeds_max` variable to `fee_rate_exceeds_max`.
3. Rephrase test log message
4.  Update `TestSimpleSend` helper transaction fee to use `MAXFEERATE`
5. Rebased to pickup cmake for easy testing!

murchandamus commented:
ACK per range-diff:

```
git range-diff 3714692644f45808a6480525abc36870aeee1de4..e5d1d25912390954b1a064c5c1f9c5d2c1726430 5acf12bafeb126f2190b3f401f95199e0eea90c9..46f0bc5b485130d3e7d14e6855d81d387950cb39
```

glozow commented on src/wallet/test/wallet_tests.cpp:
```diff
@@ -858,7 +859,8 @@ BOOST_FIXTURE_TEST_CASE(CreateWallet, TestChain100Setup)
     auto mempool_tx = TestSimpleSpend(*m_coinbase_txns[1], 0, coinbaseKey, GetScriptForRawPubKey(key.GetPubKey()));
-    BOOST_CHECK(m_node.chain->broadcastTransaction(MakeTransactionRef(mempool_tx), DEFAULT_TRANSACTION_MAXFEE, false, error));
+    auto tx_vsize = GetVirtualTransactionSize(CTransaction{mempool_tx});
+    BOOST_CHECK(m_node.chain->broadcastTransaction(MakeTransactionRef(mempool_tx), DEFAULT_TRANSACTION_MAXFEE, CFeeRate(tx_vsize, DEFAULT_TRANSACTION_MAXFEE), false, error));
```
This `CFeeRate` construction seem wrong to me - you've swapped the size and fee?

ismaelsadeeq replied:
fixed, thanks

glozow commented on src/wallet/test/wallet_tests.cpp:
```diff
@@ -900,7 +902,8 @@ BOOST_FIXTURE_TEST_CASE(CreateWallet, TestChain100Setup)
             mempool_tx = TestSimpleSpend(*m_coinbase_txns[3], 0, coinbaseKey, GetScriptForRawPubKey(key.GetPubKey()));
-            BOOST_CHECK(m_node.chain->broadcastTransaction(MakeTransactionRef(mempool_tx), DEFAULT_TRANSACTION_MAXFEE, false, error));
+            auto tx_vsize = GetVirtualTransactionSize(CTransaction{mempool_tx});
+            BOOST_CHECK(m_node.chain->broadcastTransaction(MakeTransactionRef(mempool_tx), DEFAULT_TRANSACTION_MAXFEE, CFeeRate(tx_vsize, DEFAULT_TRANSACTION_MAXFEE), false, error));
```
same

ismaelsadeeq replied:
fixed

ismaelsadeeq commented:
Forced pushed https://github.com/bitcoin/bitcoin/compare/46f0bc5b485130d3e7d14e6855d81d387950cb39..6dec2d4fed3c6ca8ebef05988fded44ad9051a77 to fix @glozow  comments

yancyribbens commented on doc/release-notes-29278.md:
```diff
@@ -0,0 +1,7 @@
+wallet startup option
+========================
+
+- A new wallet startip option `-maxfeerate` is added.
```
```suggestion
- A new wallet startup option `-maxfeerate` is added.
```

yancyribbens commented on src/qt/walletmodel.cpp:
```diff
@@ -228,7 +228,7 @@ WalletModel::SendCoinsReturn WalletModel::prepareTransaction(WalletModelTransact
         // Reject absurdly high fee. (This can never happen because the
         // wallet never creates transactions with fee greater than
-        // m_default_max_tx_fee. This merely a belt-and-suspenders check).
+        // m_max_tx_fee. This merely a belt-and-suspenders check).
```
```suggestion
        // m_max_tx_fee. This is merely a belt-and-suspenders check).
```

ismaelsadeeq replied:
fixed

ismaelsadeeq committed:
doc: add missing verb to make sentence readable

ismaelsadeeq commented on src/wallet/init.cpp:
```diff
@@ -61,6 +61,7 @@ void WalletInit::AddWalletOptions(ArgsManager& argsman) const
     argsman.AddArg("-maxapsfee=<n>", strprintf("Spend up to this amount in additional (absolute) fees (in %s) if it allows the use of partial spend avoidance (default: %s)", CURRENCY_UNIT, FormatMoney(DEFAULT_MAX_AVOIDPARTIALSPEND_FEE)), ArgsManager::ALLOW_ANY, OptionsCategory::WALLET);
     argsman.AddArg("-maxtxfee=<amt>", strprintf("Maximum total fees (in %s) to use in a single wallet transaction; setting this too low may abort large transactions (default: %s)",
         CURRENCY_UNIT, FormatMoney(DEFAULT_TRANSACTION_MAXFEE)), ArgsManager::ALLOW_ANY, OptionsCategory::DEBUG_TEST);
+    argsman.AddArg("-maxfeerate=<amt>", strprintf("Maximum fee rate (in %s/kvB) for a wallet transactions (default: %s %s/kvB)", CURRENCY_UNIT, FormatMoney(DEFAULT_MAX_TRANSACTION_FEERATE.GetFeePerK()), CURRENCY_UNIT), ArgsManager::ALLOW_ANY, OptionsCategory::WALLET);
```
Done

furszy reviewed (commented):
(very) light review

ismaelsadeeq commented on src/wallet/interfaces.cpp:
```diff
@@ -504,7 +504,7 @@ class WalletImpl : public Wallet
     }
     OutputType getDefaultAddressType() override { return m_wallet->m_default_address_type; }
-    CAmount getDefaultMaxTxFee() override { return m_wallet->m_default_max_tx_fee; }
+    CAmount getDefaultMaxTxFee() override { return m_wallet->m_max_tx_fee; }
```
Yeah, done.

ismaelsadeeq commented on test/functional/wallet_bumpfee.py:
```diff
@@ -548,9 +548,9 @@ def test_settxfee(self, rbf_node, dest_address):
     rbf_node.settxfee(Decimal("0.00000000"))  # unset paytxfee
 
-    # check that settxfee respects -maxtxfee
-    self.restart_node(1, ['-maxtxfee=0.000025'] + self.extra_args[1])
```
> I’d be surprised if we don’t have a test for that already.

Yes we do :)

ismaelsadeeq commented on src/wallet/wallet.cpp:
```diff
@@ -3046,21 +3046,42 @@ std::shared_ptr<CWallet> CWallet::Create(WalletContext& context, const std::stri
             warnings.push_back(strprintf(_("%s is set very high! Fees this large could be paid on a single transaction."), "-maxtxfee"));
+
+        // Wallet prevents creating transactions with fee rates lower than minrelaytxfee.
+        // Also the wallet prevents creating transaction with base fee above maxtxfee.
```
Fixed.

ismaelsadeeq commented:
Rebased and fixed C.I  [f94314eae4..92ea905](https://github.com/bitcoin/bitcoin/compare/ef00ee9b41dde3ccc37a0a865bc967f94314eae4..92ea90598fb00746b2e92bb2c280838e36ad0c46)

achow101 commented on src/qt/psbtoperationsdialog.cpp:
```diff
@@ -119,8 +120,9 @@ void PSBTOperationsDialog::broadcastTransaction()
 
     CTransactionRef tx = MakeTransactionRef(mtx);
     std::string err_string;
+    auto default_feerate = CFeeRate(DEFAULT_MAX_RAW_TX_FEE_RATE.GetFeePerK());
```
In f184c09015a728482f560d615ca65df1cb6a8012 "[node]: update `BroadcastTransaction` to check fee rate limit"

We should really be using the wallet's limits here if there is one rather than the defaults.

achow101 reviewed (commented):
ACK 85ebfc485bd46f061b19902dc4ccb2dbf58bd5fe

furszy commented on src/interfaces/chain.h:
```diff
@@ -199,12 +199,14 @@ class Chain
                                       const CAmount& max_tx_fee,
+                                      const CFeeRate& max_tx_fee_rate,
                                       node::TxBroadcast broadcast_method,
                                       std::string& err_string) = 0;
```
nano nit:
It would be nice to introduce a `Restrictions` struct so we can add new limitations with default values without expanding the function signature.

Also, we could return `util::Result<void>` to avoid passing `err_string` as an argument.

ismaelsadeeq replied:
I did not take the need here; it can be done in a follow-up, though.

furszy reviewed (commented):
light code review ACK 85ebfc485bd46f061b19902dc4ccb2dbf58bd5fe

ismaelsadeeq commented:
Rebased to fix merge conflict.

C.I failure in https://github.com/bitcoin/bitcoin/actions/runs/26634202425/job/78492906366?pr=29278 is unrelated.

polespinasa commented on src/wallet/wallet.cpp:
```diff
@@ -3053,7 +3053,16 @@ bool CWallet::LoadWalletArgs(std::shared_ptr<CWallet> wallet, const WalletContex
             return false;
         } else if (max_tx_fee.value() > HIGH_MAX_TX_FEE) {
             warnings.push_back(strprintf(_("%s is set very high! Fees this large could be paid on a single transaction."), "-maxtxfee"));
+        } else if (chain && CFeeRate(max_tx_fee.value(), 1000) < chain->relayMinFee()) {
```
in ebb0978ec2c1237405252ff551b4e481fada9903 [wallet]: warn user if 1kvb tx with `-maxtxfee` base fee has fee rate less than `minrelaytxfee`

Shouldn't this be an `if` instead of an `else if`?
If the previous condition is hit we are skipping this second check.

polespinasa commented on src/wallet/wallet.h:
```diff
@@ -134,10 +134,15 @@ static const bool DEFAULT_DISABLE_WALLET = false;
 //! -maxtxfee default
 constexpr CAmount DEFAULT_TRANSACTION_MAXFEE{COIN / 10};
+//! maxfeerate default
+const CFeeRate DEFAULT_MAX_TRANSACTION_FEERATE(CFeeRate(COIN / 10));
```
in 7f802e60d0f020250d17ed71615b075da31f76d2 [wallet]: add `maxfeerate` wallet startup option

nit: There's is no need to double wrap `CFeeRate` objects

polespinasa commented on src/wallet/wallet.h:
```diff
@@ -134,10 +134,15 @@ static const bool DEFAULT_DISABLE_WALLET = false;
 //! -maxtxfee will warn if called with a higher fee than this amount (in satoshis)
 constexpr CAmount HIGH_MAX_TX_FEE{100 * HIGH_TX_FEE_PER_KB};
+//! A warning will be emitted if -maxfeerate is set higher than this fee rate (in satoshis per kB).
+const CFeeRate HIGH_MAX_TX_FEERATE(CFeeRate(100 * HIGH_TX_FEE_PER_KB));
```
in 7f802e60d0f020250d17ed71615b075da31f76d2 [wallet]: add `maxfeerate` wallet startup option

nit: There's is no need to double wrap `CFeeRate` objects

polespinasa commented on src/wallet/wallet.h:
```diff
@@ -134,10 +134,15 @@ static const bool DEFAULT_DISABLE_WALLET = false;
 static const bool DEFAULT_WALLETCROSSCHAIN = false;
 //! -maxtxfee default
 constexpr CAmount DEFAULT_TRANSACTION_MAXFEE{COIN / 10};
+//! maxfeerate default
```
in 7f802e60d0f020250d17ed71615b075da31f76d2 [wallet]: add `maxfeerate` wallet startup option

nit:

```suggestion
//! -maxfeerate default
```

polespinasa commented on src/wallet/wallet.cpp:
```diff
@@ -3054,14 +3054,25 @@ bool CWallet::LoadWalletArgs(std::shared_ptr<CWallet> wallet, const WalletContex
+    if (const auto arg{args.GetArg("-maxfeerate")}) {
+        std::optional<CAmount> max_tx_fee_rate = ParseMoney(*arg);
+        if (!max_tx_fee_rate) {
+            error = AmountErrMsg("maxfeerate", args.GetArg("-maxfeerate", ""));
```
in 7f802e60d0f020250d17ed71615b075da31f76d2 [wallet]: add `maxfeerate` wallet startup option

`GetArg` is redundant the first if statement already gets it :)

```suggestion
            error = AmountErrMsg("maxfeerate", *arg);
```

polespinasa commented on src/node/interfaces.cpp:
```diff
@@ -371,14 +371,9 @@ class NodeImpl : public Node
-                                    max_tx_fee,
-                                    TxBroadcast::MEMPOOL_AND_BROADCAST_TO_ALL,
-                                    /*wait_callback=*/false);
+        return BroadcastTransaction(*m_context, std::move(tx), err_string, max_tx_fee, max_tx_fee_rate, /*broadcast_method*/ TxBroadcast::MEMPOOL_AND_BROADCAST_TO_ALL, /*wait_callback=*/false);
```
in eb703acb980b6432dbfd8744dd84129d6e6bcfbb [node]: update `BroadcastTransaction` to check fee rate limit

nit:
```suggestion
        return BroadcastTransaction(*m_context, std::move(tx), err_string, max_tx_fee, max_tx_fee_rate, /*broadcast_method=*/TxBroadcast::MEMPOOL_AND_BROADCAST_TO_ALL, /*wait_callback=*/false);
```

polespinasa commented on src/interfaces/chain.h:
```diff
@@ -199,12 +199,14 @@ class Chain
     //! @param[in] tx Transaction to process.
     //! @param[in] max_tx_fee Don't add the transaction to the mempool or
     //! broadcast it if its fee is higher than this.
+    //! @param[in]  max_tx_fee_rate reject txs with fee rate higher than this (if CFeeRate(0), the fee rate is not checked)
```
in eb703ac [node]: update BroadcastTransaction to check fee rate limit

nit: there is an extra white-space after `@param[in]`

polespinasa commented on src/wallet/test/wallet_tests.cpp:
```diff
@@ -620,8 +619,9 @@ BOOST_FIXTURE_TEST_CASE(CreateWallet, TestChain100Setup)
-    BOOST_CHECK(m_node.chain->broadcastTransaction(MakeTransactionRef(mempool_tx), DEFAULT_TRANSACTION_MAXFEE, node::TxBroadcast::MEMPOOL_NO_BROADCAST, error));
-
+    auto tx_vsize = GetVirtualTransactionSize(CTransaction{mempool_tx});
+    auto default_feerate = CFeeRate(DEFAULT_TRANSACTION_MAXFEE, tx_vsize);
```
in eb703ac [node]: update BroadcastTransaction to check fee rate limit

nit:
Here you use `default_feerate` while in line 666 you use `def_max_feerate`. Is there a reason for this? Maybe use the same variable name for consistency?

polespinasa commented on test/functional/rpc_psbt.py:
```diff
@@ -70,10 +70,12 @@
-            ["-changetype=legacy"],
-            []
+            ["-changetype=legacy", "-maxfeerate=1"],
+            [ "-maxfeerate=1"]
```
in 33d693754fb4aca240057ef996601154f2d9ce23 [wallet]: enforce `-maxfeerate` on wallet transactions

nit: extra space before wallet arg.

polespinasa reviewed (commented):
Approach ACK

Left some comments.
The PR is already well worked and reviewed so I mostly have nits before acking it.

polespinasa commented:
> * Wallet will not create a transaction with fee rate above `maxfeerate` value. And you can not set wallet fee rate with `settxfee` RPC above `maxfeerate` value.

Note that after #32138 this line of the PR description is outdated.

ismaelsadeeq commented:
Forced pushed from 440508a15deeee99613d80311581629fcf81e6c7 to ecd1e5daf6f260b585a8f12d5e9891827c66ce8a [629fcf81e6c7..ecd1e5da](https://github.com/bitcoin/bitcoin/compare/440508a15deeee99613d80311581629fcf81e6c7..ecd1e5daf6f260b585a8f12d5e9891827c66ce8a)

- https://github.com/bitcoin/bitcoin/pull/29278#discussion_r3476464716, This is correct, I added a test for that.
- https://github.com/bitcoin/bitcoin/pull/29278#discussion_r3476471363, https://github.com/bitcoin/bitcoin/pull/29278#discussion_r3476471995 Indeed, fixed.
- https://github.com/bitcoin/bitcoin/pull/29278#discussion_r3476476249, Fixed
- https://github.com/bitcoin/bitcoin/pull/29278#discussion_r3476484267, Fixed
- https://github.com/bitcoin/bitcoin/pull/29278#discussion_r3476492912, Fixed
- https://github.com/bitcoin/bitcoin/pull/29278#discussion_r3476508416 Fixed here and other places.
- https://github.com/bitcoin/bitcoin/pull/29278#discussion_r3476519276, Coming back now, I also noticed this is hard to read and not dry, and not using adjusted fee rate. I abstracted it and fixed those issues.
- https://github.com/bitcoin/bitcoin/pull/29278#discussion_r3476528388 Fixed.
- https://github.com/bitcoin/bitcoin/pull/29278#discussion_r3476528388 Fixed
- https://github.com/bitcoin/bitcoin/pull/29278#issuecomment-4802684675 Fixed.

Thanks for review @polespinasa

polespinasa commented on src/wallet/init.cpp:
```diff
@@ -61,6 +61,7 @@ void WalletInit::AddWalletOptions(ArgsManager& argsman) const
     argsman.AddArg("-maxapsfee=<n>", strprintf("Spend up to this amount in additional (absolute) fees (in %s) if it allows the use of partial spend avoidance (default: %s)", CURRENCY_UNIT, FormatMoney(DEFAULT_MAX_AVOIDPARTIALSPEND_FEE)), ArgsManager::ALLOW_ANY, OptionsCategory::WALLET);
     argsman.AddArg("-maxtxfee=<amt>", strprintf("Maximum total fees (in %s) to use in a single wallet transaction; setting this too low may abort large transactions (default: %s)",
         CURRENCY_UNIT, FormatMoney(DEFAULT_TRANSACTION_MAXFEE)), ArgsManager::ALLOW_ANY, OptionsCategory::DEBUG_TEST);
+    argsman.AddArg("-maxfeerate=<amt>", strprintf("Maximum fee rate (in %s/kvB) for wallet transactions (default: %s %s/kvB)", CURRENCY_UNIT, FormatMoney(DEFAULT_MAX_TRANSACTION_FEERATE.GetFeePerK()), CURRENCY_UNIT), ArgsManager::ALLOW_ANY, OptionsCategory::WALLET);
```
in c8b78cd9842c5aee02307884233176799ccb262b [wallet]: add `maxfeerate` wallet startup option

nit: there is a double unit specification

```
  -maxfeerate=<amt>
       Maximum fee rate (in BTC/kvB) for wallet transactions (default: 0.10
       BTC/kvB)
```

Other arguments don't specify the second `BTC/kvb`:
```
  -mintxfee=<amt>
       Fee rates (in BTC/kvB) smaller than this are considered zero fee for
       transaction creation (default: 0.00001)
  -maxapsfee=<n>
       Spend up to this amount in additional (absolute) fees (in BTC) if it
       allows the use of partial spend avoidance (default: 0.00)
 -fallbackfee=<amt>
       A fee rate (in BTC/kvB) that will be used when fee estimation has
       insufficient data. 0 to entirely disable the fallbackfee feature.
       (default: 0.00)

```

polespinasa commented on src/wallet/wallet.cpp:
```diff
@@ -3053,6 +3053,16 @@ bool CWallet::LoadWalletArgs(std::shared_ptr<CWallet> wallet, const WalletContex
+            // It is likely that some transactions with fee rates greater than or equal to the minrelaytxfee will exceed maxtxfee.
+            // In such cases, the wallet won't be able to create transactions. Therefore, warn the user.
+            warnings.push_back(strprintf(_("Invalid amount for %s=<amount>: '%s' conflicts with the minimum relay transaction feerate %s. Please set a higher %s or lower %s"),
+                                         "-maxtxfee", args.GetArg("-maxtxfee", ""), chain->relayMinFee().ToString(), "-maxtxfee", "-minrelaytxfee"));
```
in bfcc63792145cf91c55e97e0011034475fa0d081 [wallet]: warn when `-maxtxfee` conflicts with `-minrelaytxfee`

I am not sure about using `args.GetArg` here. Depending on user input the `GetArg` and `chain->relayMinFee().ToString()` will have different number of 0s.
Probably should use `CFeeRate(max_tx_fee.value(), 1000)`?

polespinasa commented on doc/release-notes-29278.md:
```diff
@@ -0,0 +1,7 @@
+- A new wallet startup option `-maxfeerate` is added.
+- This option sets the upper fee rate limit for wallet transactions.
+- The wallet will not submit or broadcast transactions whose fee rate exceeds `-maxfeerate`.
+- The default is 0.10 BTC/kvB (10,000 sat/vB).
```
in 6f05410067d3c34c9b37f741bf040d4f32a2cb34 [doc]: add release notes

Probably worth mentioning the new error type introduced (`MAX_FEE_RATE_EXCEEDED`).

polespinasa commented on src/common/messages.cpp:
```diff
@@ -138,7 +138,9 @@ bilingual_str TransactionErrorString(const TransactionError err)
             return Untranslated("Mempool internal error");
         case TransactionError::MAX_FEE_EXCEEDED:
-            return Untranslated("Fee exceeds maximum configured by user (e.g. -maxtxfee, maxfeerate)");
+            return Untranslated("Fee exceeds maximum configured by user (maxtxfee)");
+        case TransactionError::MAX_FEE_RATE_EXCEEDED:
+            return Untranslated("Fee rate exceeds maximum configured by user (maxfeerate)");
```
in 2baa98f9fdf6eee148afdb2eb9051641ddb6f5bb [util]: add a new transaction error type
Missing `-` before `maxtxfee` and `maxfeerate`

ismaelsadeeq commented:
Forced pushed from 6f05410067 to  d46adaaaaa  6f05410067...d46adaaaaa

- https://github.com/bitcoin/bitcoin/pull/29278#discussion_r3520558458 Fixed 
- https://github.com/bitcoin/bitcoin/pull/29278#discussion_r3520673277 Taken, thanks
- https://github.com/bitcoin/bitcoin/pull/29278#discussion_r3520712934 Yeah, this is a user-facing update, hence added 
- https://github.com/bitcoin/bitcoin/pull/29278#discussion_r3520729752 I am not sure the hyphen is necessary see maxburnamount. I prefer it without. Hence, not taken.

polespinasa reviewed (commented):
:)

ACK d46adaaaaa756627ca28fb1088a8938fa4cea2dc

ismaelsadeeq commented:
Rebased to fix merge conflict in `src/wallet/wallet.cpp`

furszy commented on src/wallet/spend.cpp:
```diff
@@ -1417,10 +1417,15 @@ static util::Result<CreatedTransactionResult> CreateTransactionInternal(
         return util::Error{TransactionErrorString(TransactionError::MAX_FEE_EXCEEDED)};
     }
 
+    const int64_t tx_vsize{GetVirtualTransactionSize(*tx)};
```
q: shouldn't this contemplate unsigned transactions too? E.g.
```c++
const int64_t tx_vsize{sign ? GetVirtualTransactionSize(*tx) : tx_sizes.vsize};
```

ismaelsadeeq replied:
I think using `tx_sizes.vsize` may overestimate what the fee rate is.
That makes the fee rate look lower than it really is, so the wallet may allow a tx that later fails to broadcast due to `maxfeerate`.


See a similar class of issue here https://github.com/bitcoin/bitcoin/pull/29278#issuecomment-3898336639
It's better to false-reject than false accept and fail broadcast?

maflcko commented:
What is the status here? Looks like this is a feature with 200+ comments, but no recent acks, tagged for 32.x, but it missed the feature freeze?

Maybe the milestone can be removed for now?

sedited commented:
A bit annoying that this didn't make it given the number of ACKs, but will remove this from the milestone. Would be good to get this in soon for v33.

polespinasa commented on src/wallet/wallet.h:
```diff
@@ -747,7 +747,7 @@ class CWallet final : public WalletStorage, public interfaces::Chain::Notificati
     std::optional<OutputType> m_default_change_type{};
     /** Absolute maximum transaction fee (in satoshis) used by default for the wallet */
-    CAmount m_default_max_tx_fee{DEFAULT_TRANSACTION_MAXFEE};
+    CAmount m_max_tx_fee{DEFAULT_TRANSACTION_MAXFEE};
```
in e2b19181a67c9d57e29362a10790ff0e59e74b4a scripted-diff: rename `m_default_max_tx_fee` to `m_max_tx_fee`

nit:
The comment seems wrong? The commit message says: "this value is not always the default..." and here in the code comment says it is the "default for the wallet".

polespinasa commented on src/wallet/init.cpp:
```diff
@@ -61,6 +61,7 @@ void WalletInit::AddWalletOptions(ArgsManager& argsman) const
     argsman.AddArg("-maxapsfee=<n>", strprintf("Spend up to this amount in additional (absolute) fees (in %s) if it allows the use of partial spend avoidance (default: %s)", CURRENCY_UNIT, FormatMoney(DEFAULT_MAX_AVOIDPARTIALSPEND_FEE)), ArgsManager::ALLOW_ANY, OptionsCategory::WALLET);
     argsman.AddArg("-maxtxfee=<amt>", strprintf("Maximum total fees (in %s) to use in a single wallet transaction; setting this too low may abort large transactions (default: %s)",
         CURRENCY_UNIT, FormatMoney(DEFAULT_TRANSACTION_MAXFEE)), ArgsManager::ALLOW_ANY, OptionsCategory::DEBUG_TEST);
```
in c3931437c2f82661602eea1aaeb58d9a5eebd219 [wallet]: add `maxfeerate` wallet startup option

Not introduced by this PR, but why maxtxfee sits under the category `DEBUG_TEST` and not `WALLET` like the other options?

polespinasa commented on src/wallet/wallet.cpp:
```diff
@@ -3057,14 +3057,25 @@ bool CWallet::LoadWalletArgs(std::shared_ptr<CWallet> wallet, const WalletContex
+            error = AmountErrMsg("maxfeerate", *arg);
+            return false;
+        }
+        if (chain && CFeeRate(*max_tx_fee_rate) < chain->relayMinFee()) {
```
in c393143 [wallet]: add maxfeerate wallet startup option

I think this path is not covered by any test. Worth to add a small test for it?

<details>
<summary>Suggested diff</summary>

```diff
$ git diff
diff --git a/test/functional/wallet_send.py b/test/functional/wallet_send.py
index da588b9e9c..27ae8b76e1 100755
--- a/test/functional/wallet_send.py
+++ b/test/functional/wallet_send.py
@@ -211,6 +211,24 @@ class WalletSendTest(BitcoinTestFramework):
         self.restart_node(0, extra_args=['-maxfeerate=0.00001009'])
         self.nodes[0].sendtoaddress(self.nodes[0].getnewaddress(), amount=1, fee_rate=Decimal("1.009"))
 
+        self.log.info("Test that -maxfeerate set below -minrelaytxfee is rejected as a hard error.")
+        high_min_relay_fee = '0.00020000'  # 20 sat/vB
+        low_max_fee_rate = '0.0001'        # 10 sat/vB
+        # The default wallet auto-loads on restart (load_on_startup=True). Unload it first and
+        # drop it from load-on-startup; otherwise the auto-reload trips the same hard error and
+        # the node never comes up (LoadWallets calls initError and aborts startup).
+        self.nodes[0].unloadwallet(self.default_wallet_name, load_on_startup=False)
+        self.restart_node(0, extra_args=[f'-minrelaytxfee={high_min_relay_fee}', f'-maxfeerate={low_max_fee_rate}'])
+        msg = (f"Wallet creation failed. Invalid amount for -maxfeerate=<amount>: '{low_max_fee_rate}' "
+               f"(must be at least the minrelay fee of {high_min_relay_fee} BTC/kvB to prevent stuck transactions)")
+        assert_raises_rpc_error(-4, msg, self.nodes[0].createwallet, "w_maxfeerate_below_minrelay")
+        # A -maxfeerate at or above -minrelaytxfee does not trigger the error.
+        self.restart_node(0, extra_args=[f'-minrelaytxfee={high_min_relay_fee}', '-maxfeerate=0.0003'])
+        self.nodes[0].createwallet("w_maxfeerate_ok")
+        # Restore node 0 to defaults and reload the default wallet.
+        self.restart_node(0)
+        self.nodes[0].loadwallet(self.default_wallet_name)
+
 
     def run_test(self):
         self.log.info("Setup wallets...")

```

</details>

polespinasa commented:
After reviewing this PR again, I asked some clankers to take a look at the code to see if I missed something and they found a bug.

tldr: Using sendall can return `true` for a transaction that can't enter to the mempool as it can be silently paying more fees than the allowed by `-maxfeerate`.

Clankers full explanation:

```
sendall's creation-time -maxfeerate check ignores bump fees for
unconfirmed inputs, while the broadcast-time check in BroadcastTransaction
includes them.  When fee_rate is set exactly equal to -maxfeerate and the
wallet spends unconfirmed inputs with positive bump fees, the creation check
passes (equality, ignoring bump fees) but the broadcast check fails (actual
fee > max_rate.GetFee(actual_vsize)).  The error is silently swallowed by
CommitTransaction, so sendall returns {"complete": true, "txid": ...} for a
transaction that never entered the mempool.
```

I've asked the clankers to code a functional test that proves the incorrect behavior: https://github.com/polespinasa/bitcoin/commit/62fa2e2428028043c4462ba5b5431c3d139473d0


A small fix for it could be to take also into account the total_bump_fees for the checks:
```diff
$ git diff
diff --git a/src/wallet/rpc/spend.cpp b/src/wallet/rpc/spend.cpp
index 84b0f4c483..5b2435a9a0 100644
--- a/src/wallet/rpc/spend.cpp
+++ b/src/wallet/rpc/spend.cpp
@@ -1520,10 +1520,11 @@ RPCMethod sendall()
             const std::optional<CAmount> total_bump_fees{pwallet->chain().calculateCombinedBumpFee(outpoints_spent, fee_rate)};
             CAmount effective_value = total_input_value - fee_from_size - total_bump_fees.value_or(0);
 
-            if (fee_from_size > pwallet->m_max_tx_fee) {
+            const CAmount total_fee{fee_from_size + total_bump_fees.value_or(0)};
+            if (total_fee > pwallet->m_max_tx_fee) {
                 throw JSONRPCError(RPC_WALLET_ERROR, TransactionErrorString(TransactionError::MAX_FEE_EXCEEDED).original);
             }
-            if (fee_from_size > pwallet->m_max_tx_fee_rate.GetFee(tx_size.vsize)) {
+            if (total_fee > pwallet->m_max_tx_fee_rate.GetFee(tx_size.vsize)) {
                 throw JSONRPCError(RPC_WALLET_ERROR, TransactionErrorString(TransactionError::MAX_FEE_RATE_EXCEEDED).original);
             }

```

Possible a good follow-up to fix this would be to code the TODO in wallet.cpp
